pytorch_lightning/utilities/distributed.py

```diff
@@ -55,7 +55,7 @@ def wrapped_fn(*args: Any, **kwargs: Any) -> Optional[Any]:
 
 # TODO: this should be part of the cluster environment
 def _get_rank() -> int:
-    rank_keys = ("RANK", "SLURM_PROCID", "LOCAL_RANK")
+    rank_keys = ("RANK", "LOCAL_RANK", "SLURM_PROCID")
```

Comment:
This was discussed in the original PR: https://github.com/PyTorchLightning/pytorch-lightning/pull/6802#discussion_r607593905

cc @ananthsub @awaelchli for visibility

Reply:
From https://github.com/PyTorchLightning/pytorch-lightning/pull/6802#discussion_r608400925:
> RANK = torchelastic
SLURM_PROCID = slurm
LOCAL_RANK = parity with existing setup though I think it's not right
> I set local rank last as RANK and SLURM_PROCID correspond to global rank already. 

At least in my case (running a *single* slurm job from which I then start a DDP training with `pytorch-lightning`), this is not the case, as `SLURM_PROCID` is always set to 0, on all processes started through DDP. 

I'm no Slurm expert but at least on clusters shared between many people, I don't think my case is an uncommon pattern.

Reply:
@konstantinjdobler by that do you mean you don't launch all process with slurm first? And pytorch Lightning is the entity creating the processes?

Reply:
Exactly! I have to use Slurm to *provision* resources on my cluster (e.g. a `srun` command specifying 4 GPUs in a partition which spans a sub shell in which I then execute the Python training script containing pytorch-lightning code like so: `python train.py <args...> --gpus 4`. By specifying `--gpus 4` I use the regular pytorch-lightning DDP implementation. 

This pattern actually works really well in terms of code clarity (I don't need any extra code that only works on slurm clusters, e.g. I can run the exact same code on a cluster without slurm or without DDP) and also provides frictionless access to resources on a shared cluster. Especially the point of code clarity has been one of the major selling points of pytorch-lightning for me.

Reply:
@konstantinjdobler to confirm:
1. are you also setting SLURM_NTASKS? https://github.com/PyTorchLightning/pytorch-lightning/blob/948cfd24de4f64a2980395581f15544e5e37eab0/pytorch_lightning/plugins/environments/slurm_environment.py#L82-L85 
2. Are you setting LOCAL_RANK? https://github.com/PyTorchLightning/pytorch-lightning/blob/948cfd24de4f64a2980395581f15544e5e37eab0/pytorch_lightning/plugins/environments/lightning_environment.py#L43-L50

If you answered no to both of those questions, a slightly better fix would be to iterate through the different environments lightning supports, and call detect: 
https://github.com/PyTorchLightning/pytorch-lightning/blob/948cfd24de4f64a2980395581f15544e5e37eab0/pytorch_lightning/trainer/connectors/accelerator_connector.py#L810-L812

for the first one that's available, instantiate it and call `env.global_rank()` 

then we can finally resolve the TODO that's listed here and I think that should fix the bug

Reply:
Did you try forcing `Trainer(plugins=LightningEnvironment())` (import pytorch_lightning.plugins LightningEnvironment) to bypass our automatic detection of the slurm environment?

Reply:
@awaelchli i think at least one  issue here is that the rank_zero utilities are operating independently of the cluster environment

Reply:
@ananthsub Thanks for the help! Keep in mind, I am not manually setting any of these, this is done automatically either by Slurm or pytorch-lightning
> 1. are you also setting SLURM_NTASKS? 

SLURM_NTASKS is indeed set to 1 already when I start the training script

> 2. Are you setting LOCAL_RANK? 

LOCAL_RANK is appropriately set **by pytorch-lightning** in the processes launched by DDP (e.g. LOCAL_RANK is 1 in the first additional process launched by DDP and so on). I'm not manually setting it and it is not set when I first start the training script.

> If you answered no to both of those questions, a slightly better fix would be to iterate through the different environments lightning supports, and call detect:

Both are set in my case, so not sure how to go forward. At the moment, I just fix this issue for me by monkey patching pytorch-lightning with the change proposed in my pull request. I think the crux of the issue is that I am using Slurm, but **not orchestrating the multiprocessing with Slurm**, but because Slurm environment variables are available, pytorch-lightning thinks that Slurm is taking care of the multiprocessing.

Interestingly, when calling `trainer.fit` it all works out correctly again and pytorch-lightning starts the appropriate additional processes, it's just that `rank_zero` gets "initialized" incorrectly because the SLURM_PROCID variable is present in the environment.

Reply:
> This pattern actually works really well in terms of code clarity (I don't need any extra code that only works on slurm clusters, e.g. I can run the exact same code on a cluster without slurm or without DDP) and also provides frictionless access to resources on a shared cluster. Especially the point of code clarity has been one of the major selling points of pytorch-lightning for me.

Lightning has first-class support for SLURM, such you could create the processes externally (through SLURM) without any code changes. Is this a viable option for you?

The issue now is Lightning assumes the presence of certain environment variables indicate the intention for how someone is using them. That is, if someone has set SLURM_NTASKS, or SLURM_PROCID, then we assume they're set and set correctly

Reply:
I mean I could probably create the processes externally (through SLURM) but for me that defeats the purpose of using Lightning: I don't want to deal with creating any processes, I want to let Lightning handle it.

LOCAL_RANK is set by Lightning, right? The question is if there exists a scenario where LOCAL_RANK would be set when the multiprocessing is actually handled externally. If not, to me it makes sense to check LOCAL_RANK first, to avoid issues in my scenario.

Reply:
> which spawns a sub-shell in which I then execute the Python training script containing pytorch-lightning code like so: python train.py <args...> --gpus 4. By specifying --gpus 4 I use the regular pytorch-lightning DDP implementation.

So that means if you would unset the environment variable, it would work as expected too:
```
unset SLURM_NTASKS
unset SLURM_PROCID
python train.py <args...> --gpus 4
```
Then you don't need to monkey-patch Lightning as you said.

I guess we can still go ahead with the PR but I'm also concerned with the same questions as @ananthsub

Reply:
I just tested unsetting the environment variables and it indeed does work (I still prefer the monkey patch, since I don't have to mess with the environment). The problem is solved for me, since I know about it now. 

In some sense, since I opened the issue other people with the same problem can now find the solution through a google search, but these fixes go very deep into the under-the-hood behavior of lightning and are quite unintuitive for the original problem of multiple runs being created in Weights&Biases (one for each GPU). Of course, it is up to you if believe any changes are worth it to accommodate users that use Slurm only to allocate resources and let lightning handle multiprocessing.